src/range.rs

```diff
@@ -0,0 +1,355 @@
+            type Error = TryFromIntError;
+            fn try_from(value: RangeToInclusive<$lte>) -> Result<TextRange, Self::Error> {
+                let start: TextSize = 0.into();
+                let end: TextSize = value.end.into();
+                TextRange::try_from(start..=end)
+            }
+        }
+    };
+    (TryFrom<$gt:ident> for TextRange) => {
+        impl TryFrom<Range<$gt>> for TextRange {
+            type Error = <$gt as TryInto<u32>>::Error;
+            fn try_from(value: Range<$gt>) -> Result<TextRange, Self::Error> {
+                Ok(TextRange(value.start.try_into()?, value.end.try_into()?))
+            }
+        }
+        impl TryFrom<RangeInclusive<$gt>> for TextRange {
+            type Error = TryFromIntError;
+            fn try_from(value: RangeInclusive<$gt>) -> Result<TextRange, Self::Error> {
+                let (start, end) = value.into_inner();
+                let end: TextSize = end.try_into()?;
+                // This is the only way to get a TryFromIntError currently.
+                let end = end.checked_add(1).ok_or_else(|| u8::try_from(-1).unwrap_err())?;
+                Ok(TextRange(start.try_into()?, end))
+            }
+        }
+        impl TryFrom<RangeTo<$gt>> for TextRange {
+            type Error = TryFromIntError;
+            fn try_from(value: RangeTo<$gt>) -> Result<TextRange, Self::Error> {
+                Ok(TextRange(0.into(), value.end.try_into()?))
+            }
+        }
+        impl TryFrom<RangeToInclusive<$gt>> for TextRange {
+            type Error = TryFromIntError;
+            fn try_from(value: RangeToInclusive<$gt>) -> Result<TextRange, Self::Error> {
+                let start: TextSize = 0.into();
+                let end: TextSize = value.end.try_into()?;
+                TextRange::try_from(start..=end)
+            }
+        }
+    };
+    {
+        lt TextSize [$($lt:ident)*]
+        eq TextSize [$($eq:ident)*]
+        gt TextSize [$($gt:ident)*]
+        varries     [$($var:ident)*]
+    } => {
+        $(
+            // Not `From` yet because of integer type fallback. We want e.g.
+            // `TextRange::from(0)` and `range + 1` to work, and more `From`
+            // impls means that this will try (and fail) to use i32 rather
+            // than one of the unsigned integer types that actually work.
+            conversions!(TryFrom<$lt> for TextRange);
+        )*
+
+        $(
+            conversions!(From<$eq> for TextRange);
+        )*
+
+        $(
+            conversions!(TryFrom<$gt> for TextRange);
+        )*
+
+        $(
+            conversions!(TryFrom<$var> for TextRange);
+        )*
+    };
+}
+
+// FIXME: when `default impl` is usable, change to blanket impls for [Try]Into<TextSize> instead
+conversions! {
+    lt TextSize [u8 u16]
+    eq TextSize [u32 TextSize]
+    gt TextSize [u64]
+    varries     [usize]
+}
+
+impl Into<TextRange> for &'_ TextRange {
```

Comment:
Why do we need these conversions? Is there anything similar in stdlib?

Reply:
I'd rather stick to `std` pattern of just macroing all combinations of all operators.